tests/view/RoomsTest.java

```diff
@@ -0,0 +1,178 @@
+    RoomsView roomsView =
+        RoomsUseCaseFactory.create(
+            new FirebaseRoomsDataAccessObject(client),
+            new FirebaseMessageDataAccessObject(client),
+            new FirebaseUserDataAccessObject(client),
+            roomsViewModel,
+            searchController,
+            null);
+
+    RoomsState testState = buildTestState();
+    testState.setRoomUid(testState.getAvailableRooms().get(0).getUid());
+    testState.setSendMessage("Test message!");
+    System.out.println(testState.getAvailableRooms().get(0).getMessages());
+    roomsViewModel.setState(testState);
+
+    JButton sendMessageButton = roomsView.getSendMessageButton();
+    sendMessageButton.doClick();
+
+    System.out.println(roomsViewModel.getState().getSuccess());
+
+    Assert.assertTrue(roomsViewModel.getState().getSuccess().equals("Success"));
+  }
+
+  @Test
+  public void testAddUserToRoom() {
+    RoomsViewModel roomsViewModel = new RoomsViewModel();
+    OkHttpClient client = new OkHttpClient();
+    RoomsView roomsView =
+        RoomsUseCaseFactory.create(
+            new FirebaseRoomsDataAccessObject(client),
+            new FirebaseMessageDataAccessObject(client),
+            new FirebaseUserDataAccessObject(client),
+            roomsViewModel,
+            null,
+            null);
+
+    RoomsState testState = buildTestState();
+    testState.setUserToAddEmail(createDummyDisplayUser().getEmail());
+    roomsViewModel.setState(testState);
+
+    JButton addUserButton = roomsView.getAddUserButton();
+    addUserButton.doClick();
+
+    String regex =
+        "Unable to add \\s*([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,})\\s*to room:";
+
+    Pattern pattern = Pattern.compile(regex);
+    Matcher matcher = pattern.matcher(roomsViewModel.getState().getError());
+    Assert.assertTrue(matcher.find());
+  }
+
+  private RoomsState buildTestState() {
+    RoomsState state = new RoomsState();
+    state.setUser(this.createDummyUser());
+    List<Room> rooms = new ArrayList<>();
+    Room room = createDummyRoom();
+    rooms.add(room);
+    state.setAvailableRooms(rooms);
+    state.setRoomUid(room.getUid());
+    return state;
+  }
+
+  User createDummyUser() {
+    String fakeEmail =
+        String.format("testSaveUser%s@example.com", UUID.randomUUID().toString().substring(0, 10));
+    return new User(null, fakeEmail, "Dummy User", "password", LocalDateTime.now());
+  }
+
+  DisplayUser createDummyDisplayUser() {
+    return new DisplayUser("dummy@example.com", "dummyUid");
+  }
+
+  Room createDummyRoom() {
+    DisplayUser dummyDisplayUser = new DisplayUser("dummyUid", "Dummy User");
+    List<DisplayUser> users = new ArrayList<>();
+    users.add(dummyDisplayUser);
+    List<Message> messages = new ArrayList<>();
+    messages.add(new Message(Instant.now(), "Dummy message", dummyDisplayUser));
+    Room room = new Room("dummyRoomUid", "Dummy Room", users, null);
+    room.setMessages(messages);
+    return room;
+  }
```

Comment:
There are already functions that do this in `DAOTest.java`.
Maybe call those instead?

Reply:
Yes! I was going to restructure the tests file structure to do this properly, but other people are still committing tests so I left it like this for now. Should I add this change now?